src/core/Common/Instruction.ts

```diff
@@ -1,30 +1,59 @@
-import { OpcodesNames } from './Opcodes';
+import { OpcodesNames, Opcodes } from './Opcodes';
+import { FunctionalUnitType } from './FunctionalUnit';
+import { opcodeToFunctionalUnit } from './Opcodes';
 
 export class Instruction {
-    protected _id: number;
-    protected _basicBlock: number;
-    protected _opcode: number;
+    public id: number;
+    public basicBlock: number;
+    public opcode: number;
+    protected _uuid: number;
     protected _operands: number[];
     protected _operandsString: string[];
     protected _label: string;
     protected _breakPoint: boolean;
-    protected _color: string;
+
+    public get uuid(): number {
+        return this._uuid;
+    }
+
+    public get breakPoint(): boolean {
+        return this._breakPoint;
+    }
+
+    public get operands(): number[] {
+        return this._operands;
+    }
+
+    public get label(): string {
+        return this._label;
+    }
+
+    public set label(value: string) {
+        this._label = value;
+    }
+
+    public get operandsString(): string[] {
+        return this._operandsString;
+    }
 
     constructor() {
         this._breakPoint = false;
-        this._color = 'white';
         this._operands = new Array(3);
         this._operandsString = new Array(3);
     }
 
+    instantiate(from: Instruction, cycle: number) {
```

Comment:
Ups, don't know what I was thinking when I implemented this method. 
The correct form should have been `instantiate(cycle: number): Instruction`. So it would be used like this:
`let child = parent.instantiate(cycle)`

Let me know if this form looks good to you.

Reply:
What about `let newInstruction = Instruction.fromInstruction(oldInstruction);` ?

By the way, some doubts I have:
* In your example, why are the instructions named as `parent` and `child`? Do they share any traits from inheritance?
* Why is the copied Instruction UUID being modified? UUIDs are generated to be unique, and should be immutable by definition (1,2).

One more thing: public `copy/clone` methods often work the opposite way as it is implemented here. They return a copy/clone of the instance (3). This will definitely cause confusion.

(1): https://docs.oracle.com/javase/8/docs/api/java/util/UUID.html
(2): https://docs.python.org/3/library/uuid.html
(3): https://docs.python.org/3/tutorial/datastructures.html#more-on-lists

Reply:
> In your example, why are the instructions named as parent and child? Do they share any traits from inheritance?
Not in the way you think. I'm using those terms as an analogy. I'm referring that we will always have the original instructions instances, created by the code parser and without a UUID, which must never enter the superscalar (or VLIW) execution pipeline. I call those "parents". The "children" are the copies of these same instructions issued to the execution pipeline, with will have a UUID and can exist multiples copies for the same instruction.

> Why is the copied Instruction UUID being modified? UUIDs are generated to be unique, and should be immutable by definition (1,2).
Only instances have UUIDs, but no their "parents", as it is not necessary and would only complicate things up. Using the term "UUID" is a complete botched job by my part, but I really don't any other good term.

> One more thing: public copy/clone methods often work the opposite way as it is implemented here. They return a copy/clone of the instance (3). 
Currently, yes, but not in the new implementation I'm proposing.